File: packages/protocol/src/message.ts
export const MessageType = {
  Sync: 0,
  Awareness: 1,
} as const
export type MessageType = (typeof MessageType)[keyof typeof MessageType]

export interface Message {
  type: MessageType
  payload: Uint8Array
}

/** Prefixes a y-protocols sync/awareness payload with a one-byte message type. */
export function encodeMessage(type: MessageType, payload: Uint8Array): Uint8Array {
  const out = new Uint8Array(1 + payload.length)
  out[0] = type
  out.set(payload, 1)
  return out
}

export function decodeMessage(data: Uint8Array): Message {
  const type = data[0]
  if (type !== MessageType.Sync && type !== MessageType.Awareness) {
    throw new Error(`unknown message type: ${type}`)
  }
  return { type, payload: data.subarray(1) }
}

File: internal/protocol/message.go
package protocol

import "fmt"

type MessageType byte

const (
	MessageSync      MessageType = 0
	MessageAwareness MessageType = 1
)

// RoomClosed is the WebSocket close code sent to everyone in a room when its
// host leaves. Clients must not reconnect on it: the session is over.
const RoomClosed = 4001

// EncodeMessage prefixes a y-protocols sync/awareness payload with a one-byte message type.
func EncodeMessage(t MessageType, payload []byte) []byte {
	out := make([]byte, 1+len(payload))
	out[0] = byte(t)
	copy(out[1:], payload)
	return out
}

func DecodeMessage(data []byte) (MessageType, []byte, error) {
	if len(data) == 0 {
		return 0, nil, fmt.Errorf("empty message")
	}
	t := MessageType(data[0])
	if t != MessageSync && t != MessageAwareness {
		return 0, nil, fmt.Errorf("unknown message type: %d", data[0])
	}
	return t, data[1:], nil
}
